### PR TITLE
Fix playback of playlists

### PR DESCRIPTION
**Pull request by a1rwulf**
This fixes a condition that looks like a freeze of kodi when one wants to play a m3u playlist with http-streams in it.

After https://github.com/xbmc/xbmc/commit/22b1e1e4cc861c8eb4e1b36746ac70213cafb5dd a FileItem inside a playlist holds the path of the playlist as m_strPath and the item as m_strDynPath.
Whithin the CPlayList::Expand method we may create another playlist.
The check if an item within a playlist is another playlist or not, is done by checking it's mimetype.
For the reasons above we need to check m_strDynPath to get the mimetype  instead of m_strPath.
Otherwise we may treat a livestream as a playlist and read it endlessly.

## How Has This Been Tested?
Tested playback of a m3u with a normal http stream in it on linux x11.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

@FernetMenta Shall I use m_strDynPath directly or would it be better to use the getter that falls back to m_strPath if m_strDynPath is empty?

### CODE REVIEW

File: xbmc/FileItem.cpp [hunk 1]
```diff
@@ -1456,9 +1456,9 @@ void CFileItem::FillInMimeType(bool lookup /*= true*/)
       m_mimetype = "x-directory/normal";
     else if( m_pvrChannelInfoTag )
       m_mimetype = m_pvrChannelInfoTag->InputFormat();
-    else if( StringUtils::StartsWithNoCase(m_strPath, "shout://")
-          || StringUtils::StartsWithNoCase(m_strPath, "http://")
-          || StringUtils::StartsWithNoCase(m_strPath, "https://"))
+    else if( StringUtils::StartsWithNoCase(m_strDynPath, "shout://")
```
Comment (FernetMenta) [inline xbmc/FileItem.cpp]: use GetDynPath, that falls back to path, if dynpath is not set

File: xbmc/FileItem.cpp [hunk 2]
```diff
@@ -1489,7 +1489,7 @@ void CFileItem::FillInMimeType(bool lookup /*= true*/)
 
   // change protocol to mms for the following mime-type.  Allows us to create proper FileMMS.
   if( StringUtils::StartsWithNoCase(m_mimetype, "application/vnd.ms.wms-hdr.asfv1") || StringUtils::StartsWithNoCase(m_mimetype, "application/x-mms-framed") )
-    StringUtils::Replace(m_strPath, "http:", "mms:");
+    StringUtils::Replace(m_strDynPath, "http:", "mms:");
```
Comment (FernetMenta) [inline xbmc/FileItem.cpp]: same here

**Comment by DaveTBlake** [comment]
Could this, or something similar elsewhere, fix the issue with JSON player.open of video streams with  URLs with options?
See Track ticket https://trac.kodi.tv/ticket/17914#ticket

File: xbmc/FileItem.cpp (hunk 2 above)
Comment (a1rwulf) [reply xbmc/FileItem.cpp]: This is a replace I can only do `!m_strDynPath.empty() ? m_strDynPath : m_strPath` instead.
Anyway we change m_strPath here, which is actually not desired for a function that should just fill a mimetype right?
Not sure how to fix this though...

Comment (a1rwulf) [reply xbmc/FileItem.cpp]: > EDIT: disregard, m_strPath must never be changed. if m_strDynPath is empty it has to be filled with m_strPath before replace

Do we want to do this in here?
I'd prefer setting DynPath when Path is set (of course only if it's empty at this stage), what's your opinion?

**Comment by a1rwulf** [comment]
> Could this, or something similar elsewhere, fix the issue with JSON player.open of video streams with URLs with options?
See Track ticket https://trac.kodi.tv/ticket/17914#ticket

I don't think this is related.

File: xbmc/FileItem.cpp (hunk 2 above)
Comment (FernetMenta) [reply xbmc/FileItem.cpp]: Only strDynPath can be changed. if dynPath is not empty, do the replace. if dynpath is empty, copy from path and do replace.

Comment (a1rwulf) [reply xbmc/FileItem.cpp]: done

**Review by FernetMenta — APPROVED** [review]

File: xbmc/FileItem.cpp [hunk 3]
```diff
@@ -1488,8 +1488,14 @@ void CFileItem::FillInMimeType(bool lookup /*= true*/)
   }
 
   // change protocol to mms for the following mime-type.  Allows us to create proper FileMMS.
-  if( StringUtils::StartsWithNoCase(m_mimetype, "application/vnd.ms.wms-hdr.asfv1") || StringUtils::StartsWithNoCase(m_mimetype, "application/x-mms-framed") )
-    StringUtils::Replace(m_strPath, "http:", "mms:");
+  if(StringUtils::StartsWithNoCase(m_mimetype, "application/vnd.ms.wms-hdr.asfv1") || 
+     StringUtils::StartsWithNoCase(m_mimetype, "application/x-mms-framed"))
+  {
+    if (m_strDynPath.empty())
+      m_strDynPath = m_strPath;
```
Comment (ksooo) [inline xbmc/FileItem.cpp]: This looks strange. 

@FernetMenta this is imo not what you had in mind when saying: "Only strDynPath can be changed. if dynPath is not empty, do the replace. if dynpath is empty, copy from path and do replace."

Comment (ksooo) [reply xbmc/FileItem.cpp]: nvm. seems to be okay. Sorry for the noise.

**Comment by wwark** [comment]
Hi,
Do you think it might help to resolve this issue https://trac.kodi.tv/ticket/17909 or it is not related also ?
Thanks for your return.

**Comment by a1rwulf** [comment]
@wwark I'm not sure, but looks related to me.

**Comment by a1rwulf** [comment]
Jenkins build this please

**Pull request merged**